---
title: "Access Ensembl data"
author: "Nicholas Horton (nhorton@amherst.edu) and Ben Baumer"
date: "November 14, 2017"
output: 
  pdf_document:
    fig_height: 3
    fig_width: 5
  html_document:
    fig_height: 3
    fig_width: 5
  word_document:
    fig_height: 3
    fig_width: 5
---


```{r, include=FALSE}
# Don't delete this chunk if you are using the mosaic package
# This loads the mosaic and dplyr packages
require(mosaic)
```

```{r, include=FALSE}
# Some customization.  You can alter or delete as desired (if you know what you are doing).

# This changes the default colors in lattice plots.
trellis.par.set(theme=theme.mosaic())  

# knitr settings to control how R chunks work.
require(knitr)
opts_chunk$set(
  tidy=FALSE,     # display code as typed
  size="small"    # slightly smaller font for code
)
```

## Introduction

This document is intended to describe how to access data from a MySQL database using R.  It utilizes a database of 

The website at https://www.ensembl.org states:

> Ensembl is a genome browser for vertebrate genomes that supports research in comparative genomics, evolution, sequence variation and transcriptional regulation. Ensembl annotate genes, computes multiple alignments, predicts regulatory function and collects disease data. Ensembl tools include BLAST, BLAT, BioMart and the Variant Effect Predictor (VEP) for all supported species.

A relevant paper on the topic of data management and databases in R can be found at http://chance.amstat.org/2015/04/setting-the-stage.


## Accessing data from a database using SQL commands

First we demonstrate how to access data using SQL (structured query language) commands and the `dbGetQuery()` function.  We begin by setting up a connection to the database.

```{r}
library(mosaic)
library(RMySQL)
con <- dbConnect(MySQL(), host = "ensembldb.ensembl.org",
                user = "anonymous", password = "", 
                port = 3306)
```

Next a series of SQL queries can be sent to the database.  These return R dataframes.  Let's start by seeing what databases are available.

```{r}
ds <- dbGetQuery(con, "SHOW DATABASES")
dim(ds)
grep("scrofa", ds$Database, pattern=TRUE)
```

Let's focus on the `sus_scrofa_variation_79_102` database.  
```{r warning=FALSE}
dbGetQuery(con, "USE sus_scrofa_variation_79_102")
dbGetQuery(con, "SHOW TABLES")
dbGetQuery(con, "SELECT * FROM publication")
dbGetQuery(con, "SELECT * FROM source")
ds <- dbGetQuery(con, "SELECT * from Individual LIMIT 10")
ds
```

We can track down the publication which is associated with these data.

https://www.ncbi.nlm.nih.gov/pubmed/25662601

East Balkan Swine (EBS) Sus scrofa is the only aboriginal domesticated pig breed in Bulgaria and is distributed on the western coast of the Black Sea in Bulgaria. To reveal the breed's genetic characteristics, we analysed mitochondrial DNA (mtDNA) and Y chromosomal DNA sequences of EBS in Bulgaria. Nucleotide diversity (πn ) of the mtDNA control region, including two newly found haplotypes, in 54 EBS was higher (0.014 ± 0.007) compared with that of European (0.005 ± 0.003) and Asian (0.006 ± 0.003) domestic pigs and wild boar. The median-joining network based on the mtDNA control region showed that the EBS and wild boar in Bulgaria comprised mainly two major mtDNA clades, European clade E1 (61.3%) and Asian clade A (38.7%). The coexistence of two mtDNA clades in EBS in Bulgaria may be the relict of historical pig translocation. Among the Bulgarian EBS colonies, the geographical differences in distribution of two mtDNA clades (E1 and A) could be attributed to the source pig populations and/or historical crossbreeding with imported pigs. In addition, analysis of the Y chromosomal DNA sequences for the EBS revealed that all of the EBS had haplotype HY1, which is dominant in European domestic pigs.

## Accessing a database using dplyr commands

Alternatively, a connection can be made to the server by creating a series of dplyr table objects.  
```{r}
library(mosaic)      
db <- src_mysql(dbname = "sus_scrofa_variation_79_102", 
                host = "ensembldb.ensembl.org", 
                user = "anonymous",
                port = 3306,
                password="")
Allele <- tbl(db, "allele")
```

#### Let's explore this table.
```{r}
Allele %>% 
  head()
```
